runtime/compiler/control/JitDump.cpp

```diff
@@ -153,29 +176,68 @@ jitDumpRecompileWithTracing(J9VMThread *vmThread, J9Method *ramMethod, TR::Compi
 
    // The request to use a trace log gets passed to the compilation via the optimization plan. The options object
    // created before the compile is issued will use the trace log we provide to initialize IL tracing.
-   TR_OptimizationPlan *plan = TR_OptimizationPlan::alloc(optLevel);
-   if (NULL == plan)
+   TR_OptimizationPlan *plan = NULL;
+#if defined(J9VM_OPT_JITSERVER)
+   // JITServer will create a new optimization plan on the client
+   if (compInfo->getPersistentInfo()->getRemoteCompilationMode() != JITServer::SERVER)
+#endif
       {
-      j9nls_printf(PORTLIB, J9NLS_INFO | J9NLS_STDERR, J9NLS_DMP_JIT_OPTIMIZATION_PLAN);
-      return;
-      }
+      plan = TR_OptimizationPlan::alloc(optLevel);
+      if (NULL == plan)
+         {
+         j9nls_printf(PORTLIB, J9NLS_INFO | J9NLS_STDERR, J9NLS_DMP_JIT_OPTIMIZATION_PLAN);
+         return;
+         }
 
-   plan->setInsertInstrumentation(isProfilingCompile);
-   plan->setLogCompilation(jitdumpFile);
+      plan->setInsertInstrumentation(isProfilingCompile);
+      plan->setLogCompilation(jitdumpFile);
+      }
 
    trfprintf(jitdumpFile, "<recompilation>\n");
 
    // Set the VM state of the crashed thread so the diagnostic thread can use consume it
    compInfo->setVMStateOfCrashedThread(vmThread->omrVMThread->vmState);
 
    J9::JitDumpMethodDetails details(ramMethod, optionsFromOriginalCompile, isAOTBody);
-   auto rc = compilationOK;
    auto queued = false;
-   compInfo->compileMethod(vmThread, details, oldStartPC, TR_no, &rc, &queued, plan);
+   auto rc = compilationOK;
+#if defined(J9VM_OPT_JITSERVER)
+   if (compInfo->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
+      {
+      // Inform the client that the compilation thread has crashed and send it
+      // a pointer to the JitDump file, so that it can request
+      // a JitDump recompilation from the server with an updated plan and method details
+      JITServer::ServerStream *stream = TR::compInfoPT->getMethodBeingCompiled()->_stream;
+      stream->write(JITServer::MessageType::compilationThreadCrashed, jitdumpFile);
+      stream->read<JITServer::Void>();
```

Comment:
What happens if the server is slow and the client is very fast and sends the diagnostic compilation request before the server had the chance to issue `addOutOfProcessMethodToBeCompiled`? I believe the data is going to wait on the socket until is read. Maybe we should test this scenario with a 1 sec sleep time at this point for the server.

Reply:
We could also change the order and send `compilationThreadCrashed` message after the entry has been requeued.

Reply:
Actually, this wouldn't work either, since if server processes requeued entry before we receive the response from the client, it will try to read a compilation request, but will only get `JITServer::Void`.  
One thing we could do is to read nothing at JitDump thread and just read the compilation request on the diagnostic thread as a response to `compilationThreadCrashed` message.